_posts/2018-05-23-getting-foreman-search-results-into-your-puppet-manifest.md

```diff
@@ -0,0 +1,101 @@
+
+Imagine that you are managing nodes with puppet and they are in `hostgroups` like `secure_base/role`. Where `secure_base` permits to apply all common settings to all your nodes in a secured environment. And `role` permits to apply settings dedicated to the role of a node. Now, lets say you have the constraint to not use a dns. And your customer would like to call hosts in a humain way, so with names and not with IPs. I already met this use case. Fortunatly, servers were using only one ip address.
+
+So you have to manage your `/etc/hosts` with puppet or you become crazy. 
+
+If each node execute the following code :
+
+```
+  $query = foreman({foreman_user => 'apiuser',
+                    foreman_pass => 'here_a_password',
+                    item         => 'hosts',
+                    search       => 'hostgroup_fullname ~ secure_base',
+  })
+
+  $nodes = $query['results']
+
+```
+
+In `$nodes`, you got all the needed informations about others nodes. You can check my statement by adding the following peace of code :
+
+
+```
+  $nodes.each | $_node | {
+    $_thiscert = $_node['certname']
+    $_thisip = $_node['ip']
+    notify {  "node ${_thiscert}" : message => "host ${_thisip}"}
+  }
+
+```
+
+Lets to use `$nodes` in a template with a `file` resource to manage `/etc/hosts` and you got the trick.
+
+## Unknow function foreman()
+
+If you ran the code above, you probably got an error message like `Evaluation Error: Unknown function: 'foreman'`. By default, the function `foreman()` is not available.
+
+At least two possibilities to get the function into your environement .
+
+ * You installed forman with `foreman-installer`. And so, you can simply copy `/usr/share/foreman-installer/modules/foreman/lib/puppet/parser/functions/foreman.rb` into your module managing `/etc/hosts`.
+ * And otherwise, you can [download the function foreman()](https://github.com/theforeman/puppet-foreman/blob/master/lib/puppet/parser/functions/foreman.rb) and use it into your module managing `/etc/hosts`.
+
+Probably a better way is to create a module named like `foreman-libs` containing the function `foreman()`. And add this module in your list of imported modules. To be imported by theforeman, remember that a module must:
+ * contain a file `manifests/init.pp`
+ * not contain errors, like those reported by `puppet parser validate <file.pp>`.
+
+It is also possible to write your own function to query [the API](https://theforeman.org/documentation.html) in a way that fits your needs.
+
+## Take a look about code of function foreman()
+
+The code is clearly commented. Only by reading comments, it is possible to add following informations.
+
+We used option `item` with `hosts`. But, it may be `environments`, `fact_values`, `hosts`, `hostgroups`, `puppetclasses`, `smart_proxies` or `subnets`.
+
+We didn't used option `per_page`. But, by default the function returns only 20 items in a row.
```

Comment:
didn't used -> didn't use 
No need to add 'But, ' before explaining the default behavior